external-crates/move/crates/move-vm-runtime/src/profiling/counters.rs

```diff
@@ -0,0 +1,429 @@
+
+        entries.sort_by(|a, b| b.1.cmp(&a.1));
+        entries
+    }
+
+    /// Format as a human-readable report.
+    pub fn format_report(&self) -> String {
+        let total = self.total();
+        if total == 0 {
+            return "No bytecode executions recorded.".to_string();
+        }
+
+        let mut report = String::new();
+        report.push_str(&format!("Total instructions: {}\n\n", total));
+        report.push_str("Opcode                          Count         %\n");
+        report.push_str("------------------------------------------------\n");
+
+        for (opcode, count) in self.sorted_by_frequency() {
+            let pct = (count as f64 / total as f64) * 100.0;
+            report.push_str(&format!("{:<30} {:>12} {:>8.2}%\n", format!("{:?}", opcode), count, pct));
+        }
+
+        report
+    }
+}
+
+impl Default for BytecodeSnapshot {
+    fn default() -> Self {
+        Self {
+            counts: [0u64; BYTECODE_COUNT],
+        }
+    }
+}
+
+/// Environment variable name for the profile output file path.
+pub const SUI_PROFILE_FILE_ENV: &str = "SUI_PROFILE_FILE";
+
+/// Default profile output file name.
+pub const DEFAULT_PROFILE_FILE: &str = "sui-profile.profraw";
+
+impl BytecodeSnapshot {
+    /// Format as CSV with header row.
+    /// Format: opcode,count,percentage
+    pub fn format_csv(&self) -> String {
+        let total = self.total();
+        let mut csv = String::new();
+        csv.push_str("opcode,count,percentage\n");
+
+        for (opcode, count) in self.sorted_by_frequency() {
+            let pct = if total > 0 {
+                (count as f64 / total as f64) * 100.0
+            } else {
+                0.0
+            };
+            csv.push_str(&format!("{:?},{},{:.4}\n", opcode, count, pct));
+        }
+
+        csv
+    }
+}
+
+/// Dump profile information to a file.
+///
+/// Takes a snapshot of the current bytecode counters and writes them to a CSV file.
+/// The file path is determined by:
+/// 1. The `SUI_PROFILE_FILE` environment variable if set
+/// 2. Otherwise, defaults to "sui-profile.profraw"
+///
+/// Returns `Ok(())` if the file was written successfully, or an error message.
+pub fn dump_profile_info() -> Result<(), String> {
+    let file_path =
+        std::env::var(SUI_PROFILE_FILE_ENV).unwrap_or_else(|_| DEFAULT_PROFILE_FILE.to_string());
+    dump_profile_info_to_file(&file_path)
+}
+
+/// Dump profile information to a specified file path.
+///
+/// Takes a snapshot of the current bytecode counters and writes them to a CSV file
+/// at the specified path.
+///
+/// Returns `Ok(())` if the file was written successfully, or an error message.
+pub fn dump_profile_info_to_file(file_path: &str) -> Result<(), String> {
+    let snapshot = BYTECODE_COUNTERS.snapshot();
+    let csv_content = snapshot.format_csv();
+
+    std::fs::write(file_path, csv_content)
+        .map_err(|e| format!("Failed to write profile to {}: {}", file_path, e))
+}
```

Comment:
This mechanism seems worse than reporting it alongside telemetry to me. Cc @tzakian for thoughts about that.